specs/core/1_shard-data-chains.md

```diff
@@ -299,78 +331,201 @@ def compute_crosslink_data_root(blocks: Sequence[ShardBlock]) -> Hash:
+
+    # Period transitions
+    if (state.slot + 1) % (SHARD_SLOTS_PER_BEACON_SLOT * SLOTS_PER_EPOCH * EPOCHS_PER_SHARD_PERIOD) == 0:
+        epoch = compute_epoch_of_shard_slot(state.slot)
+        earlier_committee = get_period_committee(
+            beacon_state,
+            get_shard_period_start_epoch(epoch, lookback=2),
+            state.shard,
         )
-        assert parent_block is not None
-        assert parent_block.core.slot < candidate.core.slot
-        parent_beacon_block_slot = compute_start_slot_of_epoch(compute_epoch_of_shard_slot(parent_block.core.slot))
-        assert signing_root(beacon_blocks[parent_beacon_block_slot]) == parent_block.core.beacon_chain_root
+        later_committee = get_period_committee(
+            beacon_state,
+            get_shard_period_start_epoch(epoch, lookback=1),
+            state.shard,
+        )
+        state.receipt_root = hash_tree_root(List[ShardReceiptDelta, PLACEHOLDER]([
+            ShardReceiptDelta(index, state.earlier_committee_rewards[i], state.earlier_committee_fees[i])
+            for i, index in enumerate(earlier_committee)
+        ]))
+        state.earlier_committee_rewards = state.later_committee_rewards
+        state.earlier_committee_fees = state.later_committee_fees
+        state.later_committee_rewards = [REWARD_COEFFICIENT_BASE for _ in range(len(later_committee))],
+        state.later_committee_fees = [0 for _ in range(len(later_committee))],
+    else:
+        state.receipt_root = Hash()
+    state.slot += ShardSlot(1)
+```
+
+```python
+def shard_block_transition(state: ShardState, beacon_state: BeaconState, block: ShardBlock) -> None:
+    # Check slot number
+    assert block.core.slot == state.slot
+
+    # Check parent block
+    if block.core.parent_root != Hash():
+        assert block.core.parent_root == hash_tree_root(state.most_recent_block_core)
+
+    # Calculate base reward
+    total_balance = get_total_active_balance(beacon_state)
+    base_reward = Gwei(
+        REWARD_COEFFICIENT_BASE * BASE_REWARD_FACTOR // integer_squareroot(total_balance) // BASE_REWARDS_PER_EPOCH
+    )
 
     # Check attestations
-    attester_committee = get_persistent_committee(beacon_state, shard, block.core.slot)
+    attester_committee = get_persistent_committee(beacon_state, state.shard, block.core.slot)
     pubkeys = []
+    attestations = 0
+
     for i, index in enumerate(attester_committee):
         if block.core.attester_bitfield[i]:
             pubkeys.append(beacon_state.validators[index].pubkey)
-    for i in range(len(attester_committee), MAX_PERSISTENT_COMMITTEE_SIZE * 2):
-        assert block.attester_bitfield[i] is False
+            add_reward(state, beacon_state, index, base_reward)
+            attestations += 1
+
+    for i in range(len(attester_committee), MAX_PERSISTENT_COMMITTEE_SIZE):
+        assert block.core.attester_bitfield[i] is False or block.core.attester_bitfield[i] == 0  # TODO: FIX Bitvector
```

Comment:
Somehow `block.core.attester_bitfield[i]` returns `0`.

@protolambda is it the expected behavior in pyspec?

Reply:
The elements of bitfields are typed as Bit. Bits are int subtypes, since bool can't be changed. It's a trade-off to make bitlist fully consistent with other typed lists. The int output can be cast to a bool. Or you change the other operand to Bit(False). Or, if it is not already working, we change the equality function to accept a bool type when comparing with a Bit.